### DISCUSSION TITLE
Make OperatingMode into a class

### DISCUSSION THREAD

**Discussion by frutiemax** [OpenRCT2/OpenRCT2#16769] [Q&A] [answered]
There should be some way to refactor the operating mode in the same way as the painting code in the RideTypeDescriptor. When we will add more and more operating modes for new flat rides, the code would get more and more convoluted.

What I suggest is to make OperatingMode into a class, and each OperatingMode overload the necessary methods for the vehicle update logic. This way, each OperatingMode code would be independent from the other and it will be easier to keep track of what is going on for each mode at the expense of possible code duplication.

**Comment by duncanspumpkin** [comment, marked as answer]
You will quickly find that the operating mode is not a nice simple interface that you can do this with. I've tried to do this before. You're welcome to try but it is not at all easy.

**Comment by frutiemax** [comment]
I've taken a look and it is a bit of a daunting task to try isolate the existing operating modes. When adding new operating modes, it is easy to just branch before the vehicle->update call.

I believe there can be some gain in performance refactoring the existing operating modes though, but this is a very risky task, it needs a lot of testing.
